### PR TITLE
Connection Error Handling

### PR DESCRIPTION
**Pull request by leogermani**
This PR introduces an Error Handling class for the connection package. This class will catch errors when wpcom makes an xml-rpc request to the site and the request signature is invalid.

The goal is to give better feedback to the user on how to solve the problem and, when possible, self-heal.

## How it works

When an XML-RPC request with a broken signature is received, Jetpack does not respond with any useful information (for security reasons). This leaves the user and WP.COM with no information about what is broken in the connection.

This PR adds a workflow where the site with a broken connection can securely confirm the error and inform it both to WP.COM and to the user.

1. It all starts with a request to the XML-RPC endpoint. When the signature is invalid:

2. checks if the error_code triggered was triggered in the last hour. It will only process it once an hour to avoid overloads

3. Stores the error in the database as "unverified", because we don't know if this broken request actually came from WP.COM

4. Creates a nonce for this new error.

5. Encrypts the error and send it over to WP.COM

6. WP.COM receives it and will also only process the request every once in a while. It then checks it against its database and, if the error looks legit (it has the same tokens present in WP.COM database) it will send a new request to the site, informing the token generated in step 4

7. The site receives the request from WP.COM in this new REST endpoint introduced by this PR. It checks the token and "verify" the error, meaning it will now be in the list of verified errors.

8. If the error does not get verified after 3 days, it gets deleted

9. Now that the error is verified, each error code will trigger a handle in initialization that can do different things. This PR only treats a certain number of error codes that indicates the `blog_token` is broken. It will display a simple `admin_notice` with an error message.

captura-de-tela-2020-06-19-acc80s-17 29 53

#### Jetpack product discussion
p1HpG7-9q9-p2

#### Does this pull request change what data or activity we track or use?

Yes, it collects details about connection errors (broken tokens, etc)

## Testing instructions:

To test this PR, we're going to use the "broken tokens" plugin available in the `mu-plugins` folder of this repository (which works out of the box in our docker environment)

* Start with a site successfully connected to wp.com
* Go to Jetpack > Broken tokens
* Click on "Store these options" to save your valid tokens (you will need this to fix your connection later)
* Now choose an option and break the connection. For a start, choose "Set invalid blog token"
* Go to Jetpack > XML-RPC Errors
* There should not be any errors there yet. If you see any errors, delete them before you start.
* Click on the "Jetpack Debugger" link and go to the jetpack debugger. It should show an error for your site
* Go back to Jetpack > XML-RPC Errors
* Now you should see an error in the Unverified Errors list
* Check the error nonce and timestamp. copy and save it somewhere
* Reload Jetpack Debugger. It will display the same error again
* Reload XML-RPC errors page and verify that the timestamp and nonce did not change. This verifies that the gate worked. It did not process the same error twice within an hour
* (note, `JETPACK_DEV_DEBUG` must be false for this to work)

At this point, the error was sent to WP.COM and will be processed. If confirmed WP.COM will trigger a request to our new API endpoint.

For our tests, we don't need to wait. 

* Click the "Verify Error (via API)" button
* Check that the error is copied over to the Verified errors list
* Now navigate the WP admin dashboard and confirm there is an admin notice at the top of the screen informing about the connection error.
* Disconnect your site
* Confirm that the error message is gone

## Complementary tests

* Install the [Client Example](https://github.com/Automattic/client-example) plugin and test it with Jetpack deactivated (you might need to copy the connection package over to the client example `vendor/automattic` folder if symlink is not in place)
* test with different errors (breaking the connection in different ways)
** notice the admin notice will only appear for blog token errors, which means when the requested user id is 0 or malformed and the error code is one of the following:

```
case 'malformed_token':
'token_malformed':
'no_possible_tokens':
'no_valid_token':
'unknown_token':
'could_not_sign':
'invalid_token':
'could_not_sign':
'token_mismatch':
'invalid_signature':
'signature_mismatch':
```

#### Proposed changelog entry for your changes:

* adds xml-rpc error handling to the connection package

### CODE REVIEW

**Comment by mdbitz** [comment]
lint errors (Travis CI) 

/home/travis/build/Automattic/jetpack/docker/mu-plugins/jetpack-debug-helper/inc/js/xmlrpc-errors.js
   2:40  warning  'ev' is defined but never used                       no-unused-vars
   3:3   error    Unexpected var, use let or const instead             no-var
   5:3   error    Unexpected var, use let or const instead             no-var
   9:4   error    'jetpack_broken_token_xmlrpc_errors' is not defined  no-undef
  13:14  warning  'response' is defined but never used                 no-unused-vars
  17:6   error    'jetpack_broken_token_xmlrpc_errors' is not defined  no-undef
  19:17  error    'jetpack_broken_token_xmlrpc_errors' is not defined  no-undef
  22:16  error    'response' is already declared in the upper scope    no-shadow

File: docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php [hunk 1]
```diff
@@ -67,17 +92,39 @@ public function render_ui() {
...
+				<form action="<?php echo esc_attr( admin_url( 'admin-post.php' ) ); ?>" method="post">
 					<input type="hidden" name="action" value="clear_all_xmlrpc_errors">
 					<?php wp_nonce_field( 'clear-xmlrpc-errors' ); ?>
 					<h2>
 						Current Unverified Errors
 						<input type="submit" value="Clear all unverified errors" class="button button-primary">
 					</h2>
 				</form>
+				<p>
+					Unverified errors are errors that were detected but that we don't know if they are legit and came from WP.com
```
Comment (mdbitz) [inline docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php:104]: Yes this is the jetpack-debug-helper plugin mu-plugins but shouldn't we be making these translatable strings __() esc_html__() etc

File: packages/connection/src/class-error-handler.php [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+	 * @since 8.7.0
+	 *
+	 * @param array $error_array The array representation of the error as it is stored in the database.
+	 * @return bool
+	 */
+	public function send_error_to_wpcom( $error_array ) {
+
+		$blog_id = \Jetpack_Options::get_option( 'id' );
+
+		$encrypted_data = $this->encrypt_data_to_wpcom( $error_array );
```
Comment (mdbitz) [inline packages/connection/src/class-error-handler.php:309]: check the value of $encrypted_data if false there is no purpose in making the call to WP.com as the data failed to encrypt.

File: docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php (hunk 1 above)
Comment (kraftbj) [reply docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php:104]: I'm not terribly concerned with translatable strings. They won't make it to WordPress.org, so they won't hit translate.wordpress.org, so they won't be part of the Jetpack language packs.

For straight text like this, we don't need to escape it -- only if vars are thrown in.

File: docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php [hunk 3]
```diff
@@ -67,17 +92,39 @@ public function render_ui() {
 			<div id="current_xmlrpc_errors">
 
 
-				<form action="<?php echo esc_url( admin_url( 'admin-post.php' ) ); ?>" method="post">
```
Comment (kraftbj) [inline docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php:70]: Minor given this is the helper plugin, but why the change? `esc_url` should be used anyplace an URL is used, inclusive of attribute nodes:
https://developer.wordpress.org/reference/functions/esc_url/

File: docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php (hunk 1 above)
Comment (leogermani) [reply docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php:104]: agree with Kraft

File: docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php (hunk 3 above)
Comment (leogermani) [reply docker/mu-plugins/jetpack-debug-helper/inc/class-broken-token-xmlrpc.php]: I sincerely don't remember making this change. You're right. reverting this (in two places) here https://github.com/Automattic/jetpack/pull/16194/commits/d316b9211e7067680552e385858145ad90ff8ef1

File: packages/connection/src/class-error-handler.php (hunk 2 above)
Comment (leogermani) [reply packages/connection/src/class-error-handler.php:309]: right! done in https://github.com/Automattic/jetpack/pull/16194/commits/a83e7b4fda7c1184249cdfd7a5beee47cbc0cdbb

**Comment by leogermani** [comment]
Tests passing!

File: packages/connection/src/class-error-handler.php [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+
+			switch ( $error_code ) {
+				case 'malformed_token':
+				case 'token_malformed':
+				case 'no_possible_tokens':
+				case 'no_valid_token':
+				case 'unknown_token':
+				case 'could_not_sign':
+				case 'invalid_token':
+				case 'could_not_sign':
```
Comment (kbrown9) [inline packages/connection/src/class-error-handler.php:150]: This is a duplicate; there's a 'could_not_sign' case two lines above this one.

File: packages/connection/src/class-error-handler.php [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+	 */
+	public static function get_instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Keep track of a connection error that was encoutered
```
Comment (kbrown9) [inline packages/connection/src/class-error-handler.php:175]: Just a typo:

```suggestion
	 * Keep track of a connection error that was encountered
```

File: packages/connection/src/class-error-handler.php (hunk 4 above)
Comment (leogermani) [reply packages/connection/src/class-error-handler.php]: thanks! fixed in https://github.com/Automattic/jetpack/pull/16194/commits/cbcf238b6cb6f0088f0cdf5e8c763b25ad65ad1e

File: packages/connection/src/class-error-handler.php [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+		$data = $data['signature_details'];
+
+		if ( ! isset( $data['token'] ) || empty( $data['token'] ) ) {
+			return false;
+		}
+
+		$user_id = $this->get_user_id_from_token( $data['token'] );
+
+		$error_array = array(
+			'error_code'    => $error_code,
```
Comment (kbrown9) [inline packages/connection/src/class-error-handler.php:284]: It looks like this is the only use of the `$error_code` variable in this method. Could we just call `$error->get_error_code()` here?

Comment (leogermani) [reply packages/connection/src/class-error-handler.php]: yep, https://github.com/Automattic/jetpack/pull/16194/commits/33a6661d0a83b31cabae0ca6abf3e8d004d9fb55

File: packages/connection/src/class-error-handler.php [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+
+		$verified_errors = $this->garbage_collector( $verified_errors );
+
+		return $verified_errors;
+	}
+
+	/**
+	 * Removes expired errors from the array
+	 *
+	 * This method is calleb by get_stored_errors and get_verified errors and filters their result
```
Comment (kbrown9) [inline packages/connection/src/class-error-handler.php:413]: Just a typo:

```suggestion
	 * This method is called by get_stored_errors and get_verified errors and filters their result
```

File: packages/connection/src/class-error-handler.php [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+	 * @return boolean True, if option is successfully deleted. False on failure.
+	 */
+	public function delete_verified_errors() {
+		return delete_option( self::STORED_VERIFIED_ERRORS_OPTION );
+	}
+
+	/**
+	 * Gets an error based on the nonce
+	 *
+	 * Receives a nonce and finds the related error. If error is found, move it to the verified errors option.
```
Comment (kbrown9) [inline packages/connection/src/class-error-handler.php:477]: The error isn't moved to the verified options in this method.

```suggestion
	 * Receives a nonce and finds the related error.
```

**Comment by kbrown9** [comment]
What do you think of including a link to the [Disconnecting/Reconnecting support page](https://jetpack.com/support/reconnecting-reinstalling-jetpack/) in the admin notice error message?

**Comment by leogermani** [comment]
> What do you think of including a link to the [Disconnecting/Reconnecting support page](https://jetpack.com/support/reconnecting-reinstalling-jetpack/) in the admin notice error message?

I think it's a good idea!

File: packages/connection/src/class-error-handler.php [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+	 * @param \WP_Error $error the error object.
+	 * @return boolean $should_report True if gate is open and the error should be reported.
+	 */
+	public function should_report_error( \WP_Error $error ) {
+
+		if ( defined( 'JETPACK_DEV_DEBUG' ) && JETPACK_DEV_DEBUG ) {
+			return true;
+		}
+
+		$bypass_gate = apply_filters( 'jetpack_connection_bypass_error_reporting_gate', false );
```
Comment (dereksmart) [inline packages/connection/src/class-error-handler.php:210]: Need a docblock on this

Comment (leogermani) [reply packages/connection/src/class-error-handler.php:222]: done in https://github.com/Automattic/jetpack/pull/16194/commits/f02d8c78ba0f1cec4943f88770bb5bde46dec5ce
